python/ray/data/dataset.py

```diff
@@ -5059,8 +5059,17 @@ def __setstate__(self, state):
         self._current_executor = None
 
     def __del__(self):
-        if self._current_executor and ray is not None and ray.is_initialized():
-            self._current_executor.shutdown()
+        if not self._current_executor:
+            return
+
+        # When Python shuts down, `ray` might evaluate to `<module None from None>`.
+        # This value is truthy and not `None`, so we use a try-catch rather than
+        # something like `if ray is not None`. For more information, see #42382.
+        try:
+            if ray.is_initialized():
```

Comment:
we don't need to check `ray is not None` now?

Reply:
I don't think `ray` would ever be `None` (I suspect it'd always be `<module None from None>`), but I'll add it just to be safe.

Updated.